contracts/swappers/SwapperUniswapV2Lp.sol

```diff
@@ -0,0 +1,209 @@
+import "../helpers/IUniswapV2PairFull.sol";
+import '../helpers/TransferHelper.sol';
+import "../helpers/ReentrancyGuard.sol";
+import "../Auth2.sol";
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+
+/**
+ * @dev swap usdp/any uniswapv2 lp
+ */
+contract SwapperUniswapV2Lp is AbstractSwapper {
+    using SafeMath for uint;
+
+    IERC20 public immutable WETH;
+    IERC20 public immutable USDP;
+    IERC20 public immutable USDT;
+
+    ICurveWith3crvPool public immutable USDP_3CRV_POOL;
+    int128 public immutable USDP_3CRV_POOL_USDP;
+    int128 public immutable USDP_3CRV_POOL_USDT;
+
+    ICurveTricrypto2Pool public immutable TRICRYPTO2_POOL;
+    uint256 public immutable TRICRYPTO2_USDT;
+    uint256 public immutable TRICRYPTO2_WETH;
+
+
+    constructor(
+        address _vaultParameters, address _weth,  address _usdp, address _usdt,
+        address _usdp3crvPool, int128 _usdp3crvPoolUsdpIndex, int128 _usdp3crvPoolUsdtIndex,
+        address _tricrypto2Pool, uint256 _tricrypto2PoolUsdtIndex, uint256 _tricrypto2PoolWethIndex
+    ) AbstractSwapper(_vaultParameters) {
+        require(
+            _weth != address(0)
+            && _usdp != address(0)
+            && _usdt != address(0)
+            && _usdp3crvPool != address(0)
+            && _tricrypto2Pool != address(0)
+            , "Unit Protocol Swappers: ZERO_ADDRESS"
+        );
+        require(_usdp3crvPoolUsdpIndex != _usdp3crvPoolUsdtIndex && _tricrypto2PoolUsdtIndex != _tricrypto2PoolWethIndex, "Unit Protocol Swappers: INVALID_TOKENS_INDEXES");
+
+        WETH = IERC20(_weth);
+        USDP = IERC20(_usdp);
+        USDT = IERC20(_usdt);
+
+        USDP_3CRV_POOL = ICurveWith3crvPool(_usdp3crvPool);
+        USDP_3CRV_POOL_USDP = _usdp3crvPoolUsdpIndex;
+        USDP_3CRV_POOL_USDT = _usdp3crvPoolUsdtIndex;
+
+        TRICRYPTO2_POOL = ICurveTricrypto2Pool(_tricrypto2Pool);
+        TRICRYPTO2_USDT = _tricrypto2PoolUsdtIndex;
+        TRICRYPTO2_WETH = _tricrypto2PoolWethIndex;
+
+        // for usdp to weth
+        TransferHelper.safeApprove(_usdp, _usdp3crvPool, type(uint256).max);
+        TransferHelper.safeApprove(_usdt, _tricrypto2Pool, type(uint256).max);
+
+        // for weth to usdp
+        TransferHelper.safeApprove(_weth, _tricrypto2Pool, type(uint256).max);
+        TransferHelper.safeApprove(_usdt, _usdp3crvPool, type(uint256).max);
+    }
+
+    function predictAssetOut(address _asset, uint256 _usdpAmountIn) external view override returns (uint predictedAssetAmount) {
+        IUniswapV2PairFull pair = IUniswapV2PairFull(_asset);
+        (uint256 pairWethId,,) = UniswapV2Helper.getTokenInfo(pair, address(WETH));
+
+        // USDP -> USDT
+        uint usdtAmount = USDP_3CRV_POOL.get_dy_underlying(USDP_3CRV_POOL_USDP, USDP_3CRV_POOL_USDT, _usdpAmountIn);
+
+        // USDT -> WETH
+        uint wethAmount = TRICRYPTO2_POOL.get_dy(TRICRYPTO2_USDT, TRICRYPTO2_WETH, usdtAmount);
+
+        // 1/2 WETH -> LP underlying token
```

Comment:
With fees formula is more complex, see UniswapV2Helper.calcWethToSwapBeforeMint and comment within. 
Code could be run in mathworks, without registration it could be run here https://www.mathworks.com/help/symbolic/solve.html (button try this example)